src/dune_pkg/opam_solver.ml

```diff
@@ -1869,15 +1938,35 @@ let solve_lock_dir
   let pinned_package_names = Package_name.Set.of_keys pinned_packages in
   let stats_updater = Solver_stats.Updater.init () in
   let context =
-    Context.create
-      ~pinned_packages
-      ~solver_env
-      ~repos
-      ~version_preference
-      ~local_packages:
-        (Package_name.Map.map local_packages ~f:Local_package.For_solver.to_opam_file)
-      ~stats_updater
-      ~constraints
+    let rec context =
+      lazy
+        (Context.create
+           ~pinned_packages
+           ~solver_env
+           ~repos
+           ~version_preference
+           ~local_packages:local_packages'
+           ~stats_updater
+           ~constraints)
+    and local_packages' =
+      lazy
```

Comment:
All of this "laziness" is just to allow the mutually recursive definition of Context.t and the local packages which require the context to evaluate their constraints.